File: lib/seating.ts
// Direct port of SeatingManager from seating.cpp.
// The floor is a 6x8 grid: 2 = wall, 0 = open aisle, 1 = table.
// findNearestFreeSeat runs a breadth-first search from the door, so
// the first table it reaches is guaranteed to be the closest one by
// number of steps — BFS explores the grid one ring of distance at a
// time, so whichever free table it pops first is provably nearest.

import type { GridCell, SeatResult } from "./types";

const ROWS = 6;
const COLS = 8;
const ENTRANCE: [number, number] = [3, 0];

const LAYOUT: number[][] = [
  [2, 1, 1, 1, 1, 1, 1, 1],
  [2, 1, 1, 1, 1, 1, 1, 1],
  [2, 1, 1, 1, 1, 1, 1, 1],
  [2, 1, 1, 1, 1, 1, 1, 1],
  [2, 1, 1, 1, 1, 1, 1, 1],
  [2, 1, 1, 1, 1, 1, 1, 1],
];

export class SeatingManager {
  private occupied: boolean[][] = Array.from({ length: ROWS }, () =>
    Array.from({ length: COLS }, () => false)
  );

  private isWithinBounds(row: number, col: number): boolean {
    return row >= 0 && row < ROWS && col >= 0 && col < COLS;
  }

  private isTableCell(row: number, col: number): boolean {
    return this.isWithinBounds(row, col) && LAYOUT[row][col] === 1;
  }

  private isTraversable(row: number, col: number): boolean {
    if (!this.isWithinBounds(row, col)) return false;
    return LAYOUT[row][col] === 0 || LAYOUT[row][col] === 1;
  }

  findNearestFreeSeat(): SeatResult {
    const [startRow, startCol] = ENTRANCE;
    const visited: boolean[][] = Array.from({ length: ROWS }, () =>
      Array.from({ length: COLS }, () => false)
    );
    const distance: number[][] = Array.from({ length: ROWS }, () =>
      Array.from({ length: COLS }, () => -1)
    );

    const queue: Array<[number, number]> = [[startRow, startCol]];
    visited[startRow][startCol] = true;
    distance[startRow][startCol] = 0;

    const dr = [-1, 1, 0, 0];
    const dc = [0, 0, -1, 1];

    while (queue.length > 0) {
      const [row, col] = queue.shift()!;

      if (this.isTableCell(row, col) && !this.occupied[row][col]) {
        this.occupied[row][col] = true;
        return { found: true, row, col, distance: distance[row][col] };
      }

      for (let i = 0; i < 4; i++) {
        const nr = row + dr[i];
        const nc = col + dc[i];
        if (!this.isWithinBounds(nr, nc) || visited[nr][nc]) continue;
        if (!this.isTraversable(nr, nc)) continue;
        visited[nr][nc] = true;
        distance[nr][nc] = distance[row][col] + 1;
        queue.push([nr, nc]);
      }
    }

    return { found: false };
  }

  occupySeat(row: number, col: number): boolean {
    if (!this.isTableCell(row, col) || this.occupied[row][col]) return false;
    this.occupied[row][col] = true;
    return true;
  }

  freeSeat(row: number, col: number): boolean {
    if (!this.isTableCell(row, col) || !this.occupied[row][col]) return false;
    this.occupied[row][col] = false;
    return true;
  }

  getGridState(): GridCell[][] {
    const grid: GridCell[][] = [];
    for (let r = 0; r < ROWS; r++) {
      const row: GridCell[] = [];
      for (let c = 0; c < COLS; c++) {
        if (LAYOUT[r][c] === 2) row.push("wall");
        else if (LAYOUT[r][c] === 0) row.push("aisle");
        else row.push(this.occupied[r][c] ? "table_occupied" : "table_free");
      }
      grid.push(row);
    }
    return grid;
  }
}

File: legacy-cpp-backend/seating.cpp
#include "seating.h"

#include <queue>
#include <utility>

SeatingManager g_seating_manager;

void SeatingManager::initialize() {
  layout_ = {
      {2, 2, 2, 2, 2, 2, 2, 2},
      {2, 1, 1, 1, 1, 1, 1, 2},
      {2, 1, 1, 1, 1, 1, 1, 2},
      {2, 1, 1, 1, 1, 1, 1, 2},
      {2, 1, 1, 1, 1, 1, 1, 2},
      {2, 2, 2, 2, 2, 2, 2, 2},
  };
  occupied_.assign(6, std::vector<bool>(8, false));
}

bool SeatingManager::isWithinBounds(int row, int col) const {
  return row >= 0 && row < static_cast<int>(layout_.size()) && col >= 0 &&
         col < static_cast<int>(layout_[0].size());
}

bool SeatingManager::isTableCell(int row, int col) const {
  return isWithinBounds(row, col) && layout_[row][col] == 1;
}

bool SeatingManager::isTraversable(int row, int col) const {
  if (!isWithinBounds(row, col)) {
    return false;
  }
  return layout_[row][col] == 0 || layout_[row][col] == 1;
}

SeatResult SeatingManager::findNearestFreeSeat() {
  std::queue<std::pair<int, int>> q;
  std::vector<std::vector<bool>> visited(6, std::vector<bool>(8, false));
  std::vector<std::vector<int>> distance(6, std::vector<int>(8, -1));

  const int start_row = 3;
  const int start_col = 0;
  q.push({start_row, start_col});
  visited[start_row][start_col] = true;
  distance[start_row][start_col] = 0;

  const int dr[4] = {-1, 1, 0, 0};
  const int dc[4] = {0, 0, -1, 1};

  while (!q.empty()) {
    auto current = q.front();
    q.pop();

    const int row = current.first;
    const int col = current.second;

    if (isTableCell(row, col) && !occupied_[row][col]) {
      occupied_[row][col] = true;
      return {true, row, col, distance[row][col]};
    }

    for (int i = 0; i < 4; ++i) {
      int nr = row + dr[i];
      int nc = col + dc[i];
      if (!isWithinBounds(nr, nc) || visited[nr][nc]) {
        continue;
      }
      if (!isTraversable(nr, nc)) {
        continue;
      }
      visited[nr][nc] = true;
      distance[nr][nc] = distance[row][col] + 1;
      q.push({nr, nc});
    }
  }

  return {false, -1, -1, -1};
}

bool SeatingManager::occupySeat(int row, int col) {
  if (!isTableCell(row, col) || occupied_[row][col]) {
    return false;
  }
  occupied_[row][col] = true;
  return true;
}

bool SeatingManager::freeSeat(int row, int col) {
  if (!isTableCell(row, col) || !occupied_[row][col]) {
    return false;
  }
  occupied_[row][col] = false;
  return true;
}

std::vector<std::vector<std::string>> SeatingManager::getGridState() const {
  std::vector<std::vector<std::string>> grid(6, std::vector<std::string>(8));
  for (int r = 0; r < 6; ++r) {
    for (int c = 0; c < 8; ++c) {
      if (layout_[r][c] == 2) {
        grid[r][c] = "wall";
      } else if (layout_[r][c] == 0) {
        grid[r][c] = "aisle";
      } else {
        grid[r][c] = occupied_[r][c] ? "table_occupied" : "table_free";
      }
    }
  }
  return grid;
}
